lib/RouterMixin.js

```diff
@@ -83,11 +83,30 @@ RouterMixin = {
         }
 
         self._historyListener = function (e) {
+            debug('history listener invoked', e, url, self.state.route.url);
+
             if (context) {
+                var state = self.state || {};
                 var url = self._history.getUrl();
-                debug('history listener invoked', e, url, self.state.route.url);
-                if (url !== self.state.route.url) {
-                    context.executeAction(navigateAction, {type: TYPE_POPSTATE, url: url, params: (e.state && e.state.params)});
+                var route = state.route || {};
+                var onBeforeUnloadText = typeof window.onbeforeunload === 'function' ? window.onbeforeunload() : '';
+                var confirmResult = onBeforeUnloadText ? window.confirm(onBeforeUnloadText) : true;
+                var nav = route.navigate || {};
+                var navParams = nav.params || {};
+                var historyState = {
+                    params: (nav.params || {})
+                };
+                var pageTitle = navParams.pageTitle || null;
+
+                if (!confirmResult) {
+                    // Pushes the previous history state back on top to set the correct url
+                    self._history.pushState(historyState, pageTitle, (state.route.url || url));
```

Comment:
Using `url` as fallback is debatable, since it is  going to be the new url after history popstate.

Another tricky thing is about the scroll position, if user set `enableScroll` to true for RouterMixin, RouterMixin saves the scroll position in history state.  In this case of canceling navigate, I wonder if we should add the current scroll position to the `historyState` object, if `enableScroll` is true.